### PR TITLE
[SVCS-901] CAS logout clears all sessions (TGTs) that belong to the current auth user

### PR DESCRIPTION
**Pull request by cslzchen**
## Ticket

https://openscience.atlassian.net/browse/SVCS-901

## Purpose

When a user logs out of OSF, she/he logs out of CAS as well. However, CAS only clears the current CAS session while leaving other ones of the same auth user behind. For example, CAS sessions in other browsers/computers are still active. Even after OSF removes all OSF sessions for the user, these CAS sessions can automatically log the user in to OSF again.

This is not the best practice and can become a critical when a user's password is stolen or when she/he selects "Stay signed in" and forgets to log out of OSF on a public computer. This CAS session (stolen by adversaries or left on public computers) remains active for 30 days. Reseting the password only clears the OSF sessions, not the CAS ones.

When user logs out of OSF, all CAS sessions of that user must be removed. Whether all OSF sessions are cleared is out of the scope of this ticket/PR.

This ticket is related to [PLAT-1502](https://openscience.atlassian.net/browse/PLAT-1052).

## Changes

* CAS logout clears all sessions (TGTs) that belong to the auth user.
  Add an inner predicate class that selects all matching tickets from ticket registries. These tickets must be Ticket Granting Ticket, is still active and belongs to the current auth user who tries to log
out. Destroy all matching tickets except the current one which will be handled later as usual.
* Fix a bug where invalid TGT throws institution login exception.
  Invalid TGTs should neither crash the system nor provide wrong exceptions. TGTs may have been removed during the new logout process or the ticket clean-up process. The correct behavior is to return a null credential and redirect the flow to the login page unauthenticated.

## Side effects

On corner case to test is when users tries to logout from different sessions at the same time. Locally, it is works as expected but should be tested further on staging.

## QA Notes

TBD

## Deployment Notes

TBD

### CODE REVIEW

File: cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java [hunk 1]
```diff
@@ -83,34 +93,82 @@ public OpenScienceFrameworkTerminateSessionAction(
         this.institutionHandler = institutionHandler;
     }
 
+    /** A predicate class that selects all active ticket granting tickets that have the same auth principal. */
+    public static final class ActiveTicketGrantingTicketPredicate implements Predicate {
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java:97]: A customized `predicate` class that selects all active TGTs that belong to the given auth user. This predicate is used in `this.centralAuthenticationService.getTickets()`. Implementation follows Oracle's [example](https://docs.oracle.com/javase/7/docs/api/javax/sql/rowset/Predicate.html)

File: cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java [hunk 2]
```diff
@@ -83,34 +93,82 @@ public OpenScienceFrameworkTerminateSessionAction(
...
             }
+
+            // Destroy other TGTs that belong to the same authenticated user
             if (tgt != null) {
                 final Authentication auth = tgt.getAuthentication();
                 if (auth != null) {
+                    // Obtain all TGTs that belongs to the same OSF user
+                    final String principalId = auth.getPrincipal().getId();
+                    final Collection<Ticket> tickets = this.centralAuthenticationService.getTickets(
+                            new ActiveTicketGrantingTicketPredicate(principalId)
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java:159]: [CentralAuthenticationServiceImpl](https://github.com/apereo/cas/blob/4.1.x/cas-server-core/src/main/java/org/jasig/cas/CentralAuthenticationServiceImpl.java#L579) has a public method `.getTickets()` to obtain all tickets. Use the `predicated` mentioned earlier as the selector.

File: cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java [hunk 3]
```diff
@@ -338,18 +338,19 @@ protected OpenScienceFrameworkCredential constructCredential(final RequestContex
             credential.setInstitutionId(remoteUserInfo.getInstitutionId());
             return credential;
         } else if (ticketGrantingTicketId != null) {
-            // pac4j login
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:341]: Misleading comment by myself. `ticketGrantingTicketId != null` doesn't necessary mean this is `pac4j` login. It is possible that the TGT associated with the ID has been removed (during logout process or ticket cleaning up).

File: cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java [hunk 4]
```diff
@@ -338,18 +338,19 @@ protected OpenScienceFrameworkCredential constructCredential(final RequestContex
...
             final TicketGrantingTicket ticketGrantingTicket;
             try {
                 ticketGrantingTicket = centralAuthenticationService.getTicket(
                         ticketGrantingTicketId,
                         TicketGrantingTicket.class
                 );
             } catch (final InvalidTicketException e) {
-                logger.error("Invalid Ticket Granting Ticket");
-                throw new RemoteUserFailedLoginException("Invalid Ticket Granting Ticket");
```
Comment (cslzchen) [inline cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:350]: My deepest shame ... It is arguable whether to throw an exception or not. However, throwing the wrong exception is not acceptable.

**Review by cslzchen — COMMENTED** [review]
@icereval Logout behavior has changed by this PR. Tested locally that multiple logout of the same user at the same time works as expected.

**Comment by cslzchen** [comment]
cc @icereval 

Works as expected locally with an updated version of https://github.com/CenterForOpenScience/osf.io/pull/8647. Merging now.
